File: 0094-binary-tree-inorder-traversal/iterative-depth-first-search/python/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def inorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        res = []
        stack = []
        
        cur = root
        while cur or stack:
            # Go as left as possible
            while cur:
                stack.append(cur)
                cur = cur.left

            # Process the node
            cur = stack.pop()
            res.append(cur.val)

            # Go to the right subtree
            cur = cur.right
        return res

File: 0094-binary-tree-inorder-traversal/iterative-depth-first-search/java/solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> inorderTraversal(TreeNode root) {
        List<Integer> res = new ArrayList<>();
        Deque<TreeNode> stack = new ArrayDeque<>();
        TreeNode cur = root;

        while (cur != null || !stack.isEmpty()) {
            // Go as left as possible
            while (cur != null) {
                stack.push(cur);
                cur = cur.left;
            }

            // Process the node
            cur = stack.pop();
            res.add(cur.val);

            // Go to the right subtree
            cur = cur.right;
        }
        return res;
    }
}
